src/shared/data-fd-util.c

```diff
@@ -343,3 +348,50 @@ int copy_data_fd(int fd) {
 
         return fd_reopen(tmp_fd, O_RDONLY|O_CLOEXEC);
 }
+
+int memfd_clone_fd(int fd, const char *name, int mode) {
+        _cleanup_close_ int mfd = -EBADF;
+        bool ro;
+        int r;
+
+        /* Creates a clone of a regular file in a memfd. Unlike copy_data_fd() this returns strictly a memfd
+         * (and if it can't it will fail). Thus the resulting fd is seekable, and definitely reports as
+         * S_ISREG. */
+
+        assert(fd >= 0);
+        assert(name);
+        assert(IN_SET(mode & O_ACCMODE, O_RDONLY, O_RDWR));
+        assert((mode & ~(O_RDONLY|O_RDWR|O_CLOEXEC)) == 0);
+
+        ro = (mode & O_ACCMODE) == O_RDONLY;
+
+        mfd = memfd_create(name,
+                           ((FLAGS_SET(mode, O_CLOEXEC) || ro) ? MFD_CLOEXEC : 0) |
+                           (ro ? MFD_ALLOW_SEALING : 0));
+        if (mfd < 0)
+                return -errno;
+
+        r = copy_bytes(fd, mfd, UINT64_MAX, COPY_REFLINK);
```

Comment:
COPY_REFLINK can be unset here. We know it cannot work.

Reply:
that's not that obvious. First of all, memfds are all attached to the same superblock, so reflinks might actually work if shmemfs would one day support it – in case input fd is already a memfd. There's not much lost for trying.

Also I think one day the same-superblock limitations of reflinks might get removed too, because btrfs for example exposes multiple superblocks on the same fs (each subvol is one), and they are working on allowing reflinks between subvols iirc.

Hence, when that materializes then sooner or later reflinks between tmpfs/shmfs/memfd superblocks might end up being permitted too...

So I wouldn't make unnecessary restrictions here. it's not that unlikely that this works eventually.